crates/next-core/src/page_source.rs

```diff
@@ -332,16 +337,16 @@ async fn create_page_source_for_file(
                 specificity,
                 server_root,
                 pathname,
-                path_regex,
+                params_matcher.into(),
                 ssr_entry,
                 runtime_entries,
                 fallback_page,
             ),
             create_node_rendered_source(
                 specificity,
                 server_root,
-                data_pathname,
-                data_path_regex,
+                pathname,
```

Comment:
This is part 2: Next.js has an internal Regex to detect dynamic routes, which won't match `/[dynamic].json` but will match `/[dynamic]` and `/[dynamic]/`.

Reply:
This is what Next.js does: the pathname is the original URL pathname (e.g. `/[test]`), and not the data pathname (`/_next/data/development/thing.json`).